### PR TITLE
Updated tensorflow structure, and spark default builder

### PR DESCRIPTION
**Pull request by saifjsl**
## Description

Enable tensorflow allow_soft_placement for gpu based models support. Updates TF dependency.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code improvements with no or little impact
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [CONTRIBUTING](http://nlp.johnsnowlabs.com/contribute.html) page.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -88,7 +88,16 @@ lazy val utilDependencies = Seq(
   "org.slf4j" % "slf4j-api" % "1.7.25",
   "org.apache.commons" % "commons-compress" % "1.15",
   "org.tensorflow" % "tensorflow" % "1.8.0",
-  "com.amazonaws" % "aws-java-sdk-s3" % "1.11.313"
+  // Enable the following for tensorflow support
+  //"org.tensorflow" % "libtensorflow" % "1.8.0",
```
Comment (albertoandreottiATgmail) [inline build.sbt]: How did we manage to get rid of TF dependency here? Where are we getting that dependency from now?

File: build.sbt [hunk 2]
```diff
@@ -88,7 +88,16 @@ lazy val utilDependencies = Seq(
...
+  //"org.tensorflow" % "libtensorflow_jni_gpu" % "1.8.0",
+  "com.amazonaws" % "aws-java-sdk-s3" % "1.11.313" exclude("org.slf4j", "slf4j-log4j12") exclude("log4j", "log4j"),
+  "org.apache.pdfbox" % "pdfbox" % "2.0.9" exclude("org.slf4j", "slf4j-log4j12") exclude("log4j", "log4j"),
+  "net.sourceforge.tess4j" % "tess4j" % "4.0.2" exclude("org.slf4j", "slf4j-log4j12") exclude("log4j", "log4j"),
```
Comment (albertoandreottiATgmail) [inline build.sbt]: All these excludes can be replaced with only one exclude on Tesseract dependency, like this,
"net.sourceforge.tess4j" % "tess4j" % "4.0.2" exclude("org.slf4j", "slf4j-log4j12-1.7.")

File: src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowWrapper.scala [hunk 3]
```diff
@@ -96,14 +96,28 @@ object TensorflowWrapper {
...
     // 3. Read file as SavedModelBundle
-    val graphDef = Files.readAllBytes(Paths.get(folder, "saved_model.pb"))
-    val graph = new Graph()
-    graph.importGraphDef(graphDef)
-    val session = new Session(graph)
-    session.runner.addTarget("save/restore_all")
-      .feed("save/Const", t.createTensor(Paths.get(folder, "variables").toString))
-      .run()
+    val (graph, session) = if (useBundle) {
+      val model = SavedModelBundle.load(folder, tags:_*)
```
Comment (albertoandreottiATgmail) [inline src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowWrapper.scala]: Curiosity question: any particular advantage to using this? best integration with Python maybe?

File: src/main/scala/com/johnsnowlabs/nlp/HasWordEmbeddings.scala [hunk 4]
```diff
@@ -21,16 +21,20 @@ trait HasWordEmbeddings extends AutoCloseable with ParamsAndFeaturesWritable {
 
   val nDims = new IntParam(this, "nDims", "Number of embedding dimensions")
   val indexPath = new Param[String](this, "indexPath", "File that stores Index")
+  val normalizeEmbeddings = new BooleanParam(this, "normalizeEmbeddings", "whether to use embeddings of normalized tokens (if not already normalized)")
```
Comment (albertoandreottiATgmail) [inline src/main/scala/com/johnsnowlabs/nlp/HasWordEmbeddings.scala]: That parameter name brings to me the idea of normalizing the vectors(L2 norm, and the like), and not the tokens. Maybe normalizeTokens could work better?

File: build.sbt (hunk 1 above)
Comment (saifjsl) [reply build.sbt]: I didn't, it's two lines above. The libtensorflow and libtensorflow-gpu are required for gpu usage

File: build.sbt (hunk 2 above)
Comment (saifjsl) [reply build.sbt]: I don't remember making this change, but okay

File: src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowNer.scala [hunk 5]
```diff
@@ -51,7 +49,7 @@ class TensorflowNer
         .feed(wordLengthsKey, tensors.createTensor(batchInput.wordLengths))
         .feed(charIdsKey, tensors.createTensor(batchInput.charIds))
 
-        .feed(dropoutKey, tensors.createTensor(1.1f))
+        .feed(dropoutKey, tensors.createTensor(1.0f))
```
[the remark this answers is not in the document]
Comment (saifjsl) [reply src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowNer.scala]: I don't have much to say here. We do use the dropout value when validating and doing cross-validation during training. I belive this is making it up with more epochs. What do you suggest?

File: src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowWrapper.scala (hunk 3 above)
Comment (saifjsl) [reply src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowWrapper.scala]: Pros: Use Tensorflow API for loading model
Cons: There is no Save API for ModelBundle in JAVA, so we cannot train using this. Using SavedModelBundle is only useful for reading complete models trained in python. Otherwise, we need to use Aleksei's decomposed loading method to be able to save it manually during training (using python provided graph)

File: src/main/scala/com/johnsnowlabs/nlp/HasWordEmbeddings.scala (hunk 4 above)
Comment (saifjsl) [reply src/main/scala/com/johnsnowlabs/nlp/HasWordEmbeddings.scala]: This is a good point. Although we are not normalizing tokens, just normalizing the tokens for retrieveing embeddings. I'll think of something better.
